presto-orc/src/main/java/com/facebook/presto/orc/BaseOrcRecordReader.java

```diff
@@ -0,0 +1,610 @@
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.io.Closer;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+import org.joda.time.DateTimeZone;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.orc.BaseOrcRecordReader.LinearProbeRangeFinder.createTinyStripesRangeFinder;
+import static com.facebook.presto.orc.OrcDataSourceUtils.mergeAdjacentDiskRanges;
+import static com.facebook.presto.orc.OrcReader.MAX_BATCH_SIZE;
+import static com.facebook.presto.orc.OrcWriteValidation.WriteChecksumBuilder.createWriteChecksumBuilder;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Comparator.comparingLong;
+import static java.util.Objects.requireNonNull;
+
+abstract class BaseOrcRecordReader
+        implements Closeable
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BaseOrcRecordReader.class).instanceSize();
+
+    private final OrcDataSource orcDataSource;
+
+    protected final BaseStreamReader[] baseStreamReaders;
+    protected final long[] maxBytesPerCell;
+    protected long maxCombinedBytesPerRow;
+
+    protected final long totalRowCount;
+    private final long splitLength;
+    private final Set<Integer> presentColumns;
+    protected final long maxBlockBytes;
+    protected final Map<Integer, Type> includedColumns;
+    protected long currentPosition;
+    private long currentStripePosition;
+    protected int currentBatchSize;
+    protected int nextBatchSize;
+    protected int maxBatchSize = MAX_BATCH_SIZE;
+
+    protected final List<StripeInformation> stripes;
+    private final StripeReader stripeReader;
+    protected int currentStripe = -1;
+    private AggregatedMemoryContext currentStripeSystemMemoryContext;
+
+    protected final long fileRowCount;
+    private final List<Long> stripeFilePositions;
+    protected long filePosition;
+
+    private Iterator<RowGroup> rowGroups = ImmutableList.<RowGroup>of().iterator();
+    protected int currentRowGroup = -1;
+    protected int currentGroupRowCount;
+    protected int nextRowInGroup;
+
+    private final Map<String, Slice> userMetadata;
+
+    private final AggregatedMemoryContext systemMemoryUsage;
+
+    private final Optional<OrcWriteValidation> writeValidation;
+    protected final Optional<OrcWriteValidation.WriteChecksumBuilder> writeChecksumBuilder;
```

Comment:
These validation flags are not needed for selective readers. The purpose of these flags are to verify if the orc writer writes the correct data. It does not matter what reader it uses. The validation part does not have to be in the selective reader. We could move these flags to the old reader.

Reply:
@highker Eventually, once the new readers are fully-functional and robust, I'd like to delete the current readers.